src/tools/dx.8
.TH dx 8 User Manuals
.SH NAME
dx \- Context Disk Limits
.SH SYNOPSIS
\fBdx -add-path \fIxid\fB \fIpath\fB

dx -rem-path \fIxid\fB \fIpath\fB

dx -set-limit \fIxid\fB \fIpath\fB \fIsu\fB,\fIst\fB,\fIiu\fB,\fIit\fB,\fIrr\fB

dx -get-limit \fIxid\fB \fIpath\fB
\f1
.SH DESCRIPTION
This Feature requires the use of XID Tagged Files, and allows for independent Disk Limits for different contexts on a shared partition.

The number of inodes and blocks for each filesystem is accounted, if an XID-Hash was added for the Context-Filesystem combo.

Those values, including current usage, maximum and reserved space, will be shown for filesystem queries, creating the illusion that the shared filesystem has a different usage and size, for each context.
.SH OPTIONS
.TP
\fB-add-path\f1
Add directory \fIpath\f1 to context \fIxid\f1. Use -set-limit to apply a limit to \fIpath\f1.
.TP
\fB-rem-path\f1
Remove directory \fIpath\f1 from context \fIxid\f1.
.TP
\fB-set-limit\f1
Set limit for \fIpath\f1 in context \fIxid\f1.

\fIsu\f1 - space currently used

\fIst\f1 - total space available

\fIiu\f1 - inodes currently used

\fIit\f1 - total inodes available

\fIrr\f1 - reserved for root (in %)
.TP
\fB-get-limit\f1
Get limit on \fIpath\f1 for context \fIxid\f1.
.SH AUTHOR
dx was written by Benedikt Boehm <hollow@gentoo.org>

You will find the newest version of dx in the libvserver package at \fBhttp://linux-vserver.org\f1. There is also a mailing list, you will find information about it at Linux-VServer main site.
.SH BUGS
You should report them to the mailing list.
.SH SEE ALSO
\fBix(8)\f1, \fBns(8)\f1, \fBnx(8)\f1, \fBvr(8)\f1, \fBvx(8)\f1


NEWS
01 Aug 2005: First stable release of libvserver

I'm pleased to announce libvserver 0.3, the first release considered more or
less stable. This release supports the complete API of the Linux-VServer 2.0
kernel patches (libvserver.so) and provides additional helper functions such as
capability lists (libvserver-util.so). As of now, the public interface to
libvserver.so is not subject of change anymore.

An API doc is also available via doxygen, just use the provided Doxyfile or go
to http://dev.gentoo.org/~hollow/vserver/libvserver/doc/ for the latest API doc.

Have fun!


src/tools/dx.c
// Copyright 2006 Benedikt Böhm <hollow@gentoo.org>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the
// Free Software Foundation, Inc.,
// 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <inttypes.h>

#include "tools.h"
#include "vserver.h"

static const char *rcsid = "$Id: dx.c 267 2006-07-06 08:25:19Z hollow $";

static
struct option long_opts[] = {
	COMMON_LONG_OPTS
	{ "add-path",  1, 0, 0x10 },
	{ "rem-path",  1, 0, 0x11 },
	{ "set-limit", 1, 0, 0x12 },
	{ "get-limit", 1, 0, 0x13 },
	{ NULL,        0, 0, 0 },
};

static inline
void usage(int rc)
{
	printf("Usage:\n\n"
	       "dx -add-path  <xid> <path>\n"
	       "   -rem-path  <xid> <path>\n"
	       "   -set-limit <xid> <path> <su>,<st>,<iu>,<it>,<rr>\n"
	       "   -get-limit <xid> <path>\n");
	exit(rc);
}

int main(int argc, char *argv[])
{
	INIT_ARGV0
	
	int c, i;
	xid_t xid = 0;
	char *buf;
	
	/* syscall data */
	struct vx_dlimit_base base = {
		.filename = NULL,
		.flags    = 0,
	};
	
	struct vx_dlimit dlimit = {
		.filename     = NULL,
		.space_used   = CDLIM_KEEP,
		.space_total  = CDLIM_KEEP,
		.inodes_used  = CDLIM_KEEP,
		.inodes_total = CDLIM_KEEP,
		.reserved     = CDLIM_KEEP,
		.flags        = 0,
	};
	
#define CASE_GOTO(ID, P) case ID: xid = atoi(optarg); goto P; break
	
	/* parse command line */
	while (GETOPT(c)) {
		switch (c) {
			COMMON_GETOPT_CASES
			
			CASE_GOTO(0x10, addpath);
			CASE_GOTO(0x11, rempath);
			CASE_GOTO(0x12, setlimit);
			CASE_GOTO(0x13, getlimit);
			
			DEFAULT_GETOPT_CASES
		}
	}
	
#undef CASE_GOTO
	
	goto usage;
	
addpath:
	if (argc <= optind)
		goto usage;
	
	base.filename = argv[optind];
	
	if (vx_add_dlimit(xid, &base) == -1)
		perr("vx_add_dlimit");
	
	goto out;
	
rempath:
	if (argc <= optind)
		goto usage;
	
	base.filename = argv[optind];
	
	if (vx_rem_dlimit(xid, &base) == -1)
		perr("vx_rem_dlimit");
	
	goto out;
	
setlimit:
	if (argc <= optind)
		goto usage;
	
	dlimit.filename = argv[optind];
	
	i = 0;
	
	for (buf = strtok(argv[optind+1], ","); buf != NULL; i++) {
		if (strlen(buf) < 1)
			continue;
		
#define BUF2LIM(LIM) do { \
	if (strcasecmp(buf, "inf") == 0)       LIM = CDLIM_INFINITY; \
	else if (strcasecmp(buf, "keep") == 0) LIM = CDLIM_KEEP; \
	else                                   LIM = (uint32_t) atoi(buf); \
} while(0)
		
		switch (i) {
			case 0: BUF2LIM(dlimit.space_used); break;
			case 1: BUF2LIM(dlimit.space_total); break;
			case 2: BUF2LIM(dlimit.inodes_used); break;
			case 3: BUF2LIM(dlimit.inodes_total); break;
			case 4: BUF2LIM(dlimit.reserved); break;
		}
		
#undef BUF2LIM
		
		buf = strtok(NULL, ",");
	}
	
	if (i != 5)
		goto usage;
	
	if (vx_set_dlimit(xid, &dlimit) == -1)
		perr("vx_set_dlimit");
	
	goto out;
	
getlimit:
	if (argc <= optind)
		goto usage;
	
	dlimit.filename = argv[optind];
	
	if (vx_get_dlimit(xid, &dlimit) == -1)
		perr("vx_get_dlimit");
	
#define LIM2OUT(LIM, DELIM) do { \
	if (LIM == CDLIM_INFINITY) printf("%s", "inf"); \
	else printf("%" PRIu32, LIM); \
	printf("%c", DELIM); \
} while(0)
	
	LIM2OUT(dlimit.space_used,   ',');
	LIM2OUT(dlimit.space_total,  ',');
	LIM2OUT(dlimit.inodes_used,  ',');
	LIM2OUT(dlimit.inodes_total, ',');
	LIM2OUT(dlimit.reserved,     '\n');
	
#undef LIM2OUT
	
	goto out;
	
usage:
	usage(EXIT_FAILURE);

out:
	exit(EXIT_SUCCESS);
}
